[FLINK-21030][runtime] Add global failover in case of a stop-with-savepoint failure

XComp opened the pull request:
## What is the purpose of the change

In [FLINK-21030](https://issues.apache.org/jira/browse/FLINK-21030), we experienced a bug when calling stop-with-savepoint on an embarrassingly parallel job where one pipeline failed causing the job ending up in an inconsistent state. The solution was to trigger a global failover in such a case.

## Brief change log

We couldn't rely on the `ExecutionGraph`'s `terminationFuture` to complete. The solution was to monitor the `terminationFutures` of all current executions, instead. We cause a failure if at least one of them switches to a non-FINISHED state indicating that something went wrong while stopping the job.

## Verifying this change

Two tests were added:
* testing the behavior if the failure happens during savepoint creation. No explicit restart needs to be triggered. The restart is handled by the `CheckpointFailureManager`.
* testing the behavior if the failure happens after creating the savepoint. In this case, a job failover will be triggered that brings all tasks back into `RUNNING` state.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

XComp commented:
@pnowojski @rkhachatryan Could one of you review [d9330d5](https://github.com/apache/flink/pull/14847/commits/d9330d59396d3f71a4fa65cca04a625ebc492915) to get an approval for this code change from @AHeise ?

pnowojski commented:
> Could one of you review d9330d5 

d9330d5 looks weakly good to me. Weakly as from my experience it's easy to miss some corner cases in the cancelation/cleanup code and I wouldn't be surprised that it would cause some test failures.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -908,38 +909,56 @@ public void reportCheckpointMetrics(
+                                                                                        "Inconsistent execution state after stopping with savepoint. A global fail-over was triggered to recover the job %s.",
+                                                                                        jobGraph
+                                                                                                .getJobID()));
+                                                        executionGraph.failGlobal(
+                                                                inconsistentFinalStateException);
```
I'm a bit concerned about thread-safety here: the statement seems non-thread-safe but it can be executed by any thread that terminates some `Execution`.

Previously, this block didn't access any state.

Should we use `handleAsync` with `mainThreadExecutor` here?

XComp replied:
Interesting finding. It was actually like that beforehand (using `handleAsync` in the `mainThreadExecutor`). But I ran into issues in `SavepointITCase.testStopSavepointWithBoundedInputConcurrently` which failed randomly. Switching to `handle` fixed it - hence, the change.

rkhachatryan replied:
The test was unstable on master (for some other reason) so I've just replaced it with a lower-level test.
So I think `handleAsync` shouldn't have this problem now.

XComp replied:
Cool, I rebased the branch and reverted back to `handleAsync`. One question, is there a reason to keep the leftover code from the IT test you reverted? If not, I would add a cleanup hotfix to the branch to remove things like `SavepointITCase.ischeckpointcoordinatorshutdownError` and the extensions you introduced for `InfiniteTestSource`. It looks like they we're only used for `SavepointITCase.testStopSavepointWithBoundedInputConcurrently`.

rkhachatryan replied:
Good point, I'll do that!

rkhachatryan commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java:
```diff
@@ -600,6 +616,171 @@ public void testSubmitWithUnknownSavepointPath() throws Exception {
+                assertInSnapshotCreationFailure());
+    }
+
+    @Test
+    public void testStopWithSavepointFailingAfterSnapshotCreation() throws Exception {
```
This test fails with a timeout if I run it in IDE without other tests (while others don't; and also if I run it `SavepointITCase` it succeeds).

XComp replied:
Good catch. I missed running them individually. This test failure would be fixed if we apply the change you suggested above to run the error handling in the `mainThreadExecutor`.

XComp replied:
This is fixed after letting the error handling run in the `mainThreadExecutor` again.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -908,38 +909,56 @@ public void reportCheckpointMetrics(
         // will be restarted by the CheckpointCoordinatorDeActivator.
         checkpointCoordinator.stopCheckpointScheduler();
 
+        final CompletableFuture<Collection<ExecutionState>> executionGraphTerminationFuture =
+                FutureUtils.combineAll(
+                        StreamSupport.stream(
+                                        executionGraph.getAllExecutionVertices().spliterator(),
```
Just a question, we could theoretically iterate over regions instead of all vertices, here, right?

XComp replied:
I'm not that familiar with the regions concept, yet. AFAIK, regions are collections of tasks that have to run together, right? Do they share the same `ExecutionState`? I guess, @tillrohrmann could answer that question.

rmetzger replied:
I don't think we have access to the pipelined regions in SchedulerBase (they are in the SchedulingStrategy of the DefaultScheduler). Secondly, the Regions don't have termination futures.

Disclaimer: My experience with this part of the flink code base is based on reading the code the last few minutes.

XComp replied:
Yeah, it took me also some digging to find a connection. I found  `executionGraph.getSchedulingTopology().getAllPipelinedRegions()`. Not sure, though, whether this works conceptually.

rkhachatryan reviewed (commented):
Thanks for this fix @XComp and @AHeise
The code you mentioned LGTM, but I left some questions about other parts.

XComp commented:
CC @rmetzger

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -605,6 +625,414 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
+
+        // the failure handling has to happen in the same thread as the checkpoint coordination -
+        // that's why we have to instantiate a separate ThreadExecutorService here
+        final ScheduledExecutorService singleThreadExecutorService =
```
Just for the record: We have to use a dedicated `ScheduledExecutorService` since the failure handling is triggered by the [CheckpointCanceler through the abortion mechanism](https://github.com/XComp/flink/blob/35757ed3c11f32769d6d99882596f78f8b8e12b5/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java#L1982) which runs in the `CheckpointCoordinator`'s [timer Executor](https://github.com/XComp/flink/blob/35757ed3c11f32769d6d99882596f78f8b8e12b5/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java#L726).

tillrohrmann replied:
`CheckpointCoordinator` is a mess wrt its internal threading model. This is definitely out of the scope for this PR. I think we want to refactor the `CC` for more than 3 release now.

rmetzger commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java:
```diff
@@ -539,6 +555,171 @@ public void testSubmitWithUnknownSavepointPath() throws Exception {
+        env.setParallelism(1);
+        env.getConfig()
+                .setRestartStrategy(
+                        RestartStrategies.fixedDelayRestart(expectedMaximumNumberOfRestarts, 10));
```
```suggestion
                        RestartStrategies.fixedDelayRestart(expectedMaximumNumberOfRestarts, 0));
```
Let's not waste CI time ;)

rmetzger reviewed (commented):
Thanks a lot for the fix. I left some comments, most of them are questions for my understanding of the change.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -817,6 +818,126 @@ public void testStopWithSavepointFailingWithDeclinedCheckpoint() throws Exceptio
+
+        CompletableFuture.runAsync(
+                        () -> {
+                            scheduler.updateTaskExecutionState(
+                                    new TaskExecutionState(
+                                            jobGraph.getJobID(),
+                                            succeedingExecutionAttemptId,
+                                            ExecutionState.FINISHED));
```
Why does a checkpoint abortion triggers the shut down of this task?

XComp replied:
Thanks for pointing that out. The tests got refactored as part of the most recent changes.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -817,6 +818,126 @@ public void testStopWithSavepointFailingWithDeclinedCheckpoint() throws Exceptio
+                                            failingExecutionAttemptId,
+                                            ExecutionState.FAILED));
+
+                            // the restart due to failed checkpoint handling triggering a global job
+                            // fail-over
+                            assertThat(
+                                    taskRestartExecutor.getNonPeriodicScheduledTask(), hasSize(1));
```
Why does a failed checkpoint triggers a global failover? Shouldn't the aborted savepoint stop the stop-with-savepoint operation and let everything else continue normally?

XComp replied:
The aborted Checkpoint will also trigger a [job failover in case of the synchronous checkpoint](https://github.com/XComp/flink/blob/35757ed3c11f32769d6d99882596f78f8b8e12b5/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java#L1921). 

Looking at the code again I realize that declining a checkpoint due to a failure while taking a snapshot of the state and aborting the Checkpoint due to checkpoint expiration (e.g. the TaskManager not being responsive) trigger the same error handling. That brings up the question whether one test would be enough here since we're testing the same `DefaultScheduler` functionality here. It feels like we should rather add a test to `CheckpointCoordinatorTest` to verify that an expired Checkpoint leads to the job fail-over as well. Right now, only the error handling for a declined Checkpoint is test in [CheckpointCoordinatorTest.jobFailsIfInFlightSynchronousSavepointIsDiscarded](https://github.com/XComp/flink/blob/3d400072f28ac4ff37bd38ff077bd95d57302658/flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java#L2751).

tillrohrmann replied:
Ah ok, so it is something the `CheckpointCoordinator` does. I think it is ok to not test `CheckpointCoordinator` specific functionality and defer it to some unit tests for the `CC`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -817,6 +818,126 @@ public void testStopWithSavepointFailingWithDeclinedCheckpoint() throws Exceptio
+        singleThreadExecutorService.shutdown();
+        singleThreadExecutorService.awaitTermination(1, TimeUnit.SECONDS);
+
+        assertThat(scheduler.getExecutionGraph().getState(), is(JobStatus.RUNNING));
```
This access shouldn't be thread-safe. Hence it is bound to fail eventually.

tillrohrmann replied:
I think it is dangerous to first shut down things and then assume that things have been done. This couples a lot of internal details because this only works if at the time of shutting down the executor all runnables to run need to be enqueued. If a runnable enqueues something else, then this won't work anymore.

XComp replied:
I revisited this test. I'm waiting for the job to switch back to `RUNNING` state instead of shutting down the executor service, now.

XComp replied:
My above comment became obsolete with the most recent changes.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -817,6 +818,126 @@ public void testStopWithSavepointFailingWithDeclinedCheckpoint() throws Exceptio
+        }
+
+        // we have to wait for the main executor to be finished with restarting tasks
+        singleThreadExecutorService.shutdown();
+        singleThreadExecutorService.awaitTermination(1, TimeUnit.SECONDS);
```
There is a `ExecutorUtils.gracefulShutDown` method for stopping `ExecutorServices`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/SchedulerTestingUtils.java:
```diff
@@ -144,18 +144,31 @@ public static DefaultSchedulerBuilder createSchedulerBuilder(
+
+    public static void enableCheckpointing(
+            final JobGraph jobGraph,
+            StateBackend stateBackend,
+            CheckpointStorage checkpointStorage) {
```
`@Nullable` is missing.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @XComp. The change goes in a good direction. One thing I am wondering is whether we cannot encapsulate the stop-with-savepoint logic a bit more explicitly. If this were possible then we could also test the behaviour a lot easier.

One problem I've observed in some of the tests is that they depend on quite some intricate implementation details. This is a bit dangerous as these tests might break if we change the affected components later on. In general I would suggest to not rely on internal implementation details for testing as much as possible.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -908,38 +909,57 @@ public void reportCheckpointMetrics(
+                                                                                        jobGraph
+                                                                                                .getJobID()));
+                                                        executionGraph.failGlobal(
+                                                                inconsistentFinalStateException);
```
I like this idea since what we actually wanna do is to tell the `Scheduler` and not the `ExecutionGraph`.

XComp replied:
I agree, I realize that I was putting much attention into that code line again. I could have used `SchedulerNG.handleGlobalFailure` straight away. good catch. 👍

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -908,38 +909,57 @@ public void reportCheckpointMetrics(
+                                                    Set<ExecutionState> nonFinishedStates =
+                                                            extractNonFinishedStates(
+                                                                    executionStates);
+                                                    if (throwable != null) {
```
If this case cannot happen then we could add a fail fatally call in order to make this assumption explicit.

XComp replied:
@rmetzger you're right - the Execution never completes exceptionally. I removed this code path in the `StopWithSavepointOperations` refactoring.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -605,6 +625,414 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
+        assertThat(
+                "Both of the executions where notified about the aborted checkpoint.",
+                executionAttemptIdsWithAbortedCheckpoint,
+                is(Arrays.asList(succeedingExecutionAttemptId, failingExecutionAttemptId)));
```
`containsInAnyOrder` should probably be fine.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -908,38 +909,57 @@ public void reportCheckpointMetrics(
+                .thenCompose(
+                        path ->
+                                executionGraphTerminationFuture
+                                        .handleAsync(
```
Thanks for the remark, @rmetzger. I thought about it again: We cannot move the `handleAsync` out of the `thenCompose` because the error handling that is triggered by the `handleAsync` should only be called if the savepoint creation succeeds. It would be also triggered by a failure which happened during Savepoint creation that caused all executions to terminate. The `CheckpointCoordinator` is triggering a restart in case of a Savepoint failure.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -605,6 +625,414 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
+                        ExecutionState.FINISHED));
+
+        // the restarts due to local failure handling and global job fail-over are triggered
+        assertThat(taskRestartExecutor.getNonPeriodicScheduledTask(), hasSize(2));
```
That's what I wanted to make clear with the comment above. But I could try to make the comment more explicit.

tillrohrmann replied:
Maybe there is a better way for asserting that the whole job has been restarted. I agree with Robert's observation that the test includes a lot of implementation specific knowledge.

XComp replied:
`StopWithSavepointContextTest` replaces this test focussing more on the actually functionality we want to test.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -605,6 +624,289 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
+                new TaskExecutionState(
+                        jobGraph.getJobID(),
+                        succeedingExecutionAttemptId,
+                        ExecutionState.FINISHED));
```
That is a mistake on my end - it proves that I didn't fully anticipate how the `DefaultSchedulerTest` needs to be implemented. Please correct me if I'm wrong: The tasks need to be cancelled, instead, since the cancellation call is send by the `TaskManager`. That will trigger the restart of the tasks which was initiated by the `CheckpointCoordinator`.

I came up with this code version (where the state changes to `FINISHED`/`FAILED`) due to the initial use-case of FLINK-21030. `FINISHED`/`FAILED` made the test to succeed as well because both are terminal states like `CANCEL`. Any terminal state triggers the restart of the tasks because the restart precedes the completion of the related `Execution`'s [releaseFuture](https://github.com/XComp/flink/blob/7ea786ac6af7286eb147449729316418b38cdafe/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java#L378).

I fixed it by transitioning to `CANCEL` state now.

tillrohrmann replied:
Only tasks which get canceled will need to return a `CANCELED` state. If this task does not need to be restarted, then it can simply continue running.

XComp replied:
Thanks for clarification.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -605,6 +624,289 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
+                        ExecutionState.FINISHED));
+        scheduler.updateTaskExecutionState(
+                new TaskExecutionState(
+                        jobGraph.getJobID(), failingExecutionAttemptId, ExecutionState.FAILED));
```
See my [comment above](https://github.com/apache/flink/pull/14847#discussion_r576615862) for explaination.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -908,38 +909,57 @@ public void reportCheckpointMetrics(
+                                                    return JobStatus.FINISHED;
+                                                },
+                                                mainThreadExecutor)
+                                        .thenApply(jobStatus -> path))
```
I removed the tests and replaced it by `StopWithSavepointContextTest`.

XComp commented:
I followed @tillrohrmann proposal and introduced `StopWithSavepointOperations`. Its implementation `StopWithSavepointsContext` implements the stop-with-savepoint using states which makes the code easier to read. Additionally, `StopWithSavepointContextTest` replaces the `DefaultSchedulerTest` stop-with-savepoint test cases.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -605,6 +624,289 @@ public void abortPendingCheckpointsWhenRestartingTasks() throws Exception {
+                                    jobGraph.getJobID())));
+        }
+
+        assertThat(scheduler.getExecutionGraph().getState(), is(JobStatus.RUNNING));
```
The `StopWithSavepointOperations` refactoring enables us to create more focused tests. It made the stop-with-savepoint test cases of `DefaultSchedulerTest` obsolete.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java:
```diff
@@ -0,0 +1,197 @@
+                        InitialWait + " should have preceded: No unfinishedStates is set.");
+
+                if (!context.unfinishedStates.isEmpty()) {
+                    return context.terminateExceptionWithGlobalFailover(context.unfinishedStates);
```
`terminateExceptionallyWithGlobalFailover`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java:
```diff
@@ -0,0 +1,197 @@
+        },
+        Final;
+
+        public StopWithSavepointState onSavepointCreation(
+                StopWithSavepointContext context, String path, Throwable throwable) {
+            if (throwable != null) {
+                return context.terminateExceptionally(throwable);
+            }
+
+            return handleSavepointCreation(context, path);
+        }
```
I think this logic is not correct. If we reach `WaitForSavepointCreation` first, then we want to restart the job globally if there was a savepoint failure or any of the `Executions` is not in `FINISHED`.

tillrohrmann replied:
Alternatively, we should enforce that we must stick to the order `InitialWait -> WaitForJobCompletion -> Final`

XComp replied:
Unfortunately, we cannot expect the savepoint creation to complete before the `ExecutionStates` switch to a terminal state. Checkpoint abortion is triggered [after the resources of `Executions` are released](https://github.com/XComp/flink/blob/fe413286b8082c4089a05604da0aba285950563e/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java#L1181). Hence, we cannot rely on the Savepoint creation future terminating before the `Executions` terminate.

tillrohrmann replied:
But if we are sure that both signals will eventually happen, then we could order the signals before they are given to the `StopWithSavepointOperation`, right? I think by making the callbacks dependent on each other it could work.

XComp replied:
`StopWithSavepointManager` handles the order of completion now.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java:
```diff
@@ -0,0 +1,197 @@
+        WaitForJobTermination {
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
```
I am wondering whether the `context` really helps in the overall design. It seems that we use the context for storing state and to trigger some methods. Making the state an inner class could solve the latter problem. Using classes instead of enums could then allow us to store the state with the state instance instead of some fields in the outer class which are never used in the outer context.

XComp replied:
I refactored it to use classes now. 👍

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java:
```diff
@@ -0,0 +1,197 @@
+            @Override
+            public StopWithSavepointState onExecutionsTermination(
+                    StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+                context.unfinishedStates = extractUnfinishedStates(executionStates);
```
Couldn't we already trigger a global failover if `unfinishedStates.isEmpty() == false`?

XComp replied:
The code was refactored making this comment obsolete.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointOperations.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * {@code StopWithSavepointOperations} collects the steps for creating a savepoint and waiting for
+ * the job to stop.
```
What are the steps?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointOperations.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Handles the Savepoint creation termination.
+     *
+     * @param path the path to the newly created savepoint.
+     * @param throwable the {@code Throwable} in case of failure.
```
What is the contract between `path` and `throwable`?

tillrohrmann replied:
Maybe it is a bit easier to have `onSavepointSuccess(String path)` and `onSavepointFailure(Throwable throwable)`

XComp replied:
I followed your advice and added to methods for the two cases. That helps differentiating the two cases! 👍

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java:
```diff
@@ -0,0 +1,249 @@
+
+    @Before
+    public void setup() throws Exception {
+        jobGraph = new JobGraph();
+
+        final JobVertex jobVertex = new JobVertex("vertex #0");
+        jobVertex.setInvokableClass(NoOpInvokable.class);
+        jobGraph.addVertex(jobVertex);
+
+        // checkpointInterval has to be set to a value lower than Long.MAX_VALUE to enable
+        // periodic checkpointing - only then can we enable/disable the CheckpointCoordinator
+        SchedulerTestingUtils.enablePeriodicCheckpointing(jobGraph, Long.MAX_VALUE - 1);
+        scheduler =
+                SchedulerTestingUtils.createSchedulerBuilder(
+                                jobGraph, ComponentMainThreadExecutorServiceAdapter.forMainThread())
+                        .setFutureExecutor(new DirectScheduledExecutorService())
+                        .build();
+        scheduler.startScheduling();
+
+        // the checkpoint scheduler is stopped before triggering the stop-with-savepoint
+        disableCheckpointScheduler();
+
+        testInstance = new StopWithSavepointContext(jobGraph.getJobID(), scheduler, this.log);
```
This looks quite involved for a simple unit test. This could be an indicator that `StopWithSavepointContext` is not super easy to test yet.

XComp replied:
You're right. I refactored the `StopWithSavepointContextTest`. It uses `TestingSchedulerNG` now instead of `DefaultScheduler` which removes some complexity from the test code.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointOperations.java:
```diff
@@ -0,0 +1,54 @@
+     * @param path the path to the newly created savepoint.
+     * @param throwable the {@code Throwable} in case of failure.
+     */
+    void handleSavepointCreation(String path, Throwable throwable);
```
```suggestion
    void handleSavepointCreation(@Nullable String path, @Nullable Throwable throwable);
```

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java:
```diff
@@ -0,0 +1,249 @@
+                    testInstance.handleSavepointCreation("savepoint-path", null);
+                    testInstance.handleExecutionTermination(
+                            // the task failed and was restarted
+                            Collections.singletonList(ExecutionState.RUNNING));
```
`RUNNING` is not a terminal state. If an `Execution` is restarted, then the old `Execution` will reach a terminal state and then we will create a new `Execution`. Hence, we are testing something which is not possible atm I think.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java:
```diff
@@ -0,0 +1,249 @@
+
+        disableCheckpointScheduler();
+
+        stopWithSavepointCompletion.run();
+
+        // the task gets cancelled before triggering the restart
+        ExecutionAttemptID executionAttemptID =
+                scheduler
+                        .getExecutionGraph()
+                        .getAllExecutionVertices()
+                        .iterator()
+                        .next()
+                        .getCurrentExecutionAttempt()
+                        .getAttemptId();
+        scheduler.updateTaskExecutionState(
+                new TaskExecutionState(
+                        jobGraph.getJobID(), executionAttemptID, ExecutionState.CANCELED));
+
+        restartExecutor.triggerScheduledTasks();
+
+        try {
+            testInstance.getResult().get();
+            fail("An ExecutionException is expected.");
+        } catch (Throwable e) {
+            final Optional<FlinkException> expectedFlinkException =
+                    ExceptionUtils.findThrowable(e, FlinkException.class);
+            final String expectedMessage =
+                    String.format(
+                            "Inconsistent execution state after stopping with savepoint. At least one execution is still in one of the following states: RUNNING. A global fail-over is triggered to recover the job %s.",
+                            jobGraph.getJobID());
+            assertTrue(expectedFlinkException.isPresent());
+            assertThat(
+                    expectedFlinkException.get(), FlinkMatchers.containsMessage(expectedMessage));
+        }
+
+        // the global fail-over puts all tasks into DEPLOYING state again
+        assertExecutionStates(scheduler, ExecutionState.DEPLOYING);
+
+        // the CheckpointCoordinator should be enabled again
+        assertCheckpointSchedulingEnabled();
```
This looks quite complicated for a simple unit test. This is usually a sign that the subject under test is not yet super easy to test.

XComp replied:
I refactored the code accordingly introducing `CheckpointScheduling` as an interface and switching to `SchedulerNG` for the global failover triggering. I hesitated to move `SchedulerNG.handleGlobalFailure` as it felt to be at the right place in `SchedulerNG` already.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java:
```diff
@@ -0,0 +1,249 @@
+                            Collections.singletonList(ExecutionState.RUNNING));
+                    testInstance.handleSavepointCreation("savepoint-path", null);
+                });
+    }
```
What happens if `handleSavepointCreation(null, throwable)`?

XComp replied:
The contract was described properly in the new implementation. Corresponding tests were added to `StopWithSavepointTerminationHandlerImplTest`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java:
```diff
@@ -0,0 +1,249 @@
+    @Test
+    public void testSavepointCreationFailure() {
+        final Exception exception = new Exception("Expected exception during savepoint creation.");
+        testInstance.handleSavepointCreation(null, exception);
```
What happens if `testInstance.onExecutionsTermination` is called after the savepoint has failed (e.g. if an unrelated global failover is triggered at a later point in time)?

XComp replied:
This test case is covered in `StopWithSavepointTerminationHandlerImplTest.testSavepointCreationFailureWithFailingExecutions`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java:
```diff
@@ -0,0 +1,197 @@
+                    "No onSavepointCreation should have been called in " + this.name() + " state.");
+        }
+
+        public StopWithSavepointState onExecutionsTermination(
+                StopWithSavepointContext context, Collection<ExecutionState> executionStates) {
+            throw new IllegalStateException(
+                    "No onExecutionsTermination should have been called in "
+                            + this.name()
+                            + " state.");
+        }
```
I think this does not work because a savepoint can fail and a later point in time `onExecutionsTermination` could be called (e.g. if an unrelated global failover is triggered).

XComp replied:
You're right. I added all possible test cases to the most recent code version.

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @XComp. The general idea looks good to me. I had a couple of comments. Please take a look.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -835,8 +837,15 @@ public void updateAccumulators(final AccumulatorSnapshot accumulatorSnapshot) {
     }
 
-    private void startCheckpointScheduler(final CheckpointCoordinator checkpointCoordinator) {
+    @Override
```
The `CheckpointCoordinator` does not change during the lifetime of an `ExecutionGraph`. That's why, I thought of making this change to introduce the `CheckpointScheduling` interface.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -908,49 +909,38 @@ public void reportCheckpointMetrics(
-                                startCheckpointScheduler(checkpointCoordinator);
-                                throw new CompletionException(throwable);
-                            }
+        StopWithSavepointContext stopWithSavepointContext =
```
I renamed the class into `StopWithSavepointOperationsImpl`.

XComp commented:
The following state diagram reflect the most recent code changes to support the code review.

![stop-with-savepoint-states](https://user-images.githubusercontent.com/1101012/108392692-3f7c7c00-7213-11eb-82c8-edfafd86d1da.png)

XComp commented:
Adapting the `StopWithSavepointOperations` interface in [46f8c9e](https://github.com/apache/flink/commit/46f8c9e) helped reducing the complexity of the state graph:
![stop-with-savepoint-states-46f8c9e](https://user-images.githubusercontent.com/1101012/108512327-54afe400-72c1-11eb-95e7-e7a696782bf1.png)

tillrohrmann commented:
Nice :-)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointScheduling.java:
```diff
@@ -0,0 +1,31 @@
+package org.apache.flink.runtime.checkpoint;
+
+/**
+ * {@code CheckpointScheduling} provides methods for starting and stopping the periodic scheduling.
```
```suggestion
 * {@code CheckpointScheduling} provides methods for starting and stopping the periodic scheduling of checkpoints.
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+    private State state = new WaitingForSavepoint();
+
+    public <S extends SchedulerNG & CheckpointScheduling> StopWithSavepointTerminationHandlerImpl(
+            @Nonnull JobID jobId, @Nonnull S schedulerWithCheckpointing, @Nonnull Logger log) {
```
The convention in Flink is that everything which is not annotated with `@Nullable` is considered to be non-null: https://flink.apache.org/contributing/code-style-and-quality-common.html#nullability-of-the-mutable-parts. Hence, `@Nonnull` should not be necessary here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+                        mainThreadExecutor)
+                .thenCompose(
+                        aVoid ->
+                                terminatedExecutionsFuture.thenAcceptAsync(
```
Why is `async` here required?

XComp replied:
It's not. Correct me, if I'm wrong: The execution will happen in the main thread if we use a non-Async operation (like the `thenRun` you suggested) since the previous operation is forced to run in the main thread already through `handleAsync`

tillrohrmann replied:
Yes exactly.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+            CompletableFuture<CompletedCheckpoint> completedSavepointFuture,
+            CompletableFuture<Collection<ExecutionState>> terminatedExecutionsFuture,
+            ComponentMainThreadExecutor mainThreadExecutor) {
+        completedSavepointFuture
+                .whenCompleteAsync(
+                        (completedSavepoint, throwable) -> {
+                            if (throwable != null) {
+                                handleSavepointCreationFailure(throwable);
+                            } else {
+                                handleSavepointCreation(completedSavepoint);
+                            }
+                        },
+                        mainThreadExecutor)
+                .thenCompose(
+                        aVoid ->
+                                terminatedExecutionsFuture.thenAcceptAsync(
+                                        this::handleExecutionsTermination, mainThreadExecutor));
```
I think this logic does not really belong to this class. If the class states that one first needs to signal the savepoint status and then only the termination status, then it should be good enough. Moreover, by not doing this part in this class, the class does not need to know about the concept of a main thread.

XComp replied:
Good point. I refactored it and introduced a `StopWithSavepointTerminationManager` to enforce the right execution order.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+                            }
+                        },
+                        mainThreadExecutor)
+                .thenCompose(
```
`thenRun` should be fine here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+                        mainThreadExecutor)
+                .thenCompose(
+                        aVoid ->
+                                terminatedExecutionsFuture.thenAcceptAsync(
```
Maybe guard this result via `FutureUtils.assertNoException` in order to make sure that `terminatedExecutionsFuture` can never be completed exceptionally.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+
+        scheduler.handleGlobalFailure(inconsistentFinalStateException);
+
+        try {
+            completedSavepoint.discard();
+        } catch (Exception e) {
+            log.warn(
+                    "Error occurred while cleaning up completed savepoint due to stop-with-savepoint failure.",
+                    e);
+            inconsistentFinalStateException.addSuppressed(e);
+        }
```
This can entail IO operations. IO operations should not be executed in the main thread because they can be blocking. If we want to discard the savepoint, then we should pass in an `ioExecutor` which can be used for this.

tillrohrmann replied:
Alternatively, we need to give the `completedSavepoint` to the owner of this class in case of a termination failure. The owner would then be responsible for cleaning the savepoint up.

XComp replied:
👍 I left it in `StpWithSavepointTerminationHandlerImpl` because it's part of the actual handle logic and therefore is easier to be tested there.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+            CompletableFuture<Collection<ExecutionState>> terminatedExecutionsFuture,
+            ComponentMainThreadExecutor mainThreadExecutor) {
+        completedSavepointFuture
+                .whenCompleteAsync(
+                        (completedSavepoint, throwable) -> {
+                            if (throwable != null) {
+                                handleSavepointCreationFailure(throwable);
+                            } else {
+                                handleSavepointCreation(completedSavepoint);
+                            }
+                        },
+                        mainThreadExecutor)
```
I would suggest to consume an exception via `handleAsync` and then add a `ExceptionUtils.assertNoException` in order to make potential misusages of the state call visible. Otherwise if `handleSavepoint*` fails, then this exception will be swallowed by the returned future.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+            final Set<ExecutionState> unfinishedStates = extractUnfinishedStates(executionStates);
+
+            if (unfinishedStates.isEmpty()) {
+                terminateSuccessfully(completedSavepoint.getExternalPointer());
+                return new FinalState();
+            }
+
+            terminateExceptionallyWithGlobalFailover(completedSavepoint, unfinishedStates);
+            return new FinalState();
+        }
+    }
+
+    private final class FinalState implements State {
+
+        @Override
+        public State onExecutionsTermination(Collection<ExecutionState> executionStates) {
+            return this;
+        }
+    }
+
+    private interface State {
+
+        default State onSavepointCreation(CompletedCheckpoint completedSavepoint) {
+            throw new UnsupportedOperationException(
+                    this.getClass().getSimpleName()
+                            + " state does not support onSavepointCreation.");
+        }
+
+        default State onSavepointCreationFailure(Throwable throwable) {
+            throw new UnsupportedOperationException(
+                    this.getClass().getSimpleName()
+                            + " state does not support onSavepointCreationFailure.");
+        }
+
+        default State onExecutionsTermination(Collection<ExecutionState> executionStates) {
+            throw new UnsupportedOperationException(
+                    this.getClass().getSimpleName()
+                            + " state does not support onExecutionsTermination.");
+        }
+    }
```
This looks simple and good to understand :-)

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImplTest.java:
```diff
@@ -0,0 +1,330 @@
+    public void testSavepointCreationFailureAfterSuccessfulTermination() {
+        assertSavepointCreationFailure(
+                (expectedException, completedSavepointFuture, executionsTerminatedFuture) -> {
+                    executionsTerminatedFuture.complete(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                    completedSavepointFuture.completeExceptionally(expectedException);
```
If we make the ordering of signals the responsibility of a different component, then we wouldn't have this combinatorial explosion of test cases. Then we would have test cases for the different signal order completions and test cases for the `StopWithSavepointTerminationHandlerImpl` where we can concentrate on the handler specific functionality.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImplTest.java:
```diff
@@ -0,0 +1,330 @@
+        // the checkpoint scheduling should be enabled in case of failure
+        assertTrue("Checkpoint scheduling should be enabled.", checkpointScheduling.isEnabled());
+
+        assertTrue("The savepoint should be cleaned up.", streamStateHandle.isDisposed());
```
How do we assert that we call the global failover?

XComp replied:
Good catch! 👍

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,258 @@
+        return result;
+    }
+
+    private synchronized void handleSavepointCreation(CompletedCheckpoint completedCheckpoint) {
+        final State oldState = state;
+        state = state.onSavepointCreation(completedCheckpoint);
+
+        log.debug(
+                "Stop-with-savepoint transitioned from {} to {} on savepoint creation handling for job {}.",
+                oldState,
+                state,
+                jobId);
+    }
+
+    private synchronized void handleSavepointCreationFailure(Throwable throwable) {
+        final State oldState = state;
+        state = state.onSavepointCreationFailure(throwable);
+
+        log.debug(
+                "Stop-with-savepoint transitioned from {} to {} on savepoint creation failure handling for job {}.",
+                oldState,
+                state,
+                jobId);
+    }
+
+    private synchronized void handleExecutionsTermination(
```
Why are these methods `synchronized`?

XComp replied:
True, this was there to cover the concurrent calls of savepoint creation and execution termination. It's not necessary anymore having the order of calls as a contract for the class. I'm gonna remove them

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @XComp. I think the changes go in a very good direction. I had a couple of comments and questions. Please take a look.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -835,8 +839,15 @@ public void updateAccumulators(final AccumulatorSnapshot accumulatorSnapshot) {
-    private void startCheckpointScheduler(final CheckpointCoordinator checkpointCoordinator) {
+    @Override
+    public void stopCheckpointScheduler() {
+        getCheckpointCoordinator().stopCheckpointScheduler();
```
`getCheckpointCoordinator` can return `null` if checkpointing is not enabled. Maybe add a precondition that checkpointing is enabled.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandler.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * {@code StopWithSavepointTerminationHandler} handles the termination both steps necessary for the
```
```suggestion
 * {@code StopWithSavepointTerminationHandler} handles the termination steps necessary for the
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandler.java:
```diff
@@ -0,0 +1,71 @@
+public interface StopWithSavepointTerminationHandler {
+
+    /**
+     * Returns the a {@code CompletableFuture} referring to the result of the stop-with-savepoint
```
```suggestion
     * Returns a {@code CompletableFuture} referring to the result of the stop-with-savepoint
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    @Override
+    public void handleSavepointCreation(
```
Just out of curiosity why did you choose this API call as the entry into the `StopWithSavepointTerminationHandler` compared to exposing `handleSavepointCreationFailure` and `handleSavepointCreationSuccess`?

XComp replied:
`StopWithSavepointTerminationHandler` and `StopWithSavepointTerminationManager` are separated in a way that the manager focuses solely on the order of the completion whereas the handler deals with the result of the termination. Having `StopWithSavepointTerminationHandler` offer two different methods for success and failure case would mean that we have to put the logic of differentiating these two into the manager. That would destroy the separation of responsibilities partially. 
As a consequence, we would have to distinguish between the failure case and the success case in the manager's test class again.

XComp replied:
I realized that I refactor the corresponding `StopWithSavepointTerminationManagerTest` accordingly. We don't need all the different test cases there. I'm fixing this now.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,297 @@
+    private void discardSavepoint(
+            CompletedCheckpoint completedSavepoint,
+            FlinkException inconsistentFinalStateException) {
+        CompletableFuture.runAsync(
```
Why did you choose to create a `CompletableFuture` here?

XComp replied:
You got me here. I thought, that `CompletableFuture.runAsync` could be used interchangeably. I changed it to `Executor.execute`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,297 @@
+    private void discardSavepoint(
+            CompletedCheckpoint completedSavepoint,
+            FlinkException inconsistentFinalStateException) {
+        CompletableFuture.runAsync(
+                () -> {
+                    try {
+                        completedSavepoint.discard();
+                    } catch (Exception e) {
+                        log.warn(
+                                "Error occurred while cleaning up completed savepoint due to stop-with-savepoint failure.",
+                                e);
+                        inconsistentFinalStateException.addSuppressed(e);
+                    }
+                },
+                ioExecutor);
```
How are we ensuring that this action is executed before we shut down the scheduler?

XComp replied:
The `ioExecutor` of the `Scheduler` is passed down from the `JobManagerSharedServices.scheduledExecutorService` which has a [shutdown procedure](https://github.com/XComp/flink/blob/2c4e0ab921ccfaf003073ee50faeae4d4e4f4c93/flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobManagerSharedServices.java#L88) for gracefully shutting down the `Executor`. Shouldn't this be independent from the `Scheduler` shutdown? 🤔

tillrohrmann replied:
Well, it depends on whose responsibility it is to clean up the savepoint. The existing shutdown procedure won't guarantee that enqueued tasks are executed. `ExecutorService.shutdownNow` will only execute the currently running tasks and clear all other tasks. Hence, there is no guarantee that the savepoint is actually removed.

XComp replied:
I went for the easy fix to add a graceful shutdown to the `JobManagerSharedServices` for now after discussing solutions with @zentol today. Here's what we came up with:
The `StopWithSavepointTerminationHandler` interface would need to be adapted: Instead of providing a dedicated method for returning the result, `StopWithSavepointTerminationHandler.handleExecutionsTermination` has to return the `CompletableFuture` result. This enables us to add an optional execution for discarding the savepoint in case of global fail-over. The `StopWithSavepointTerminationManager` would forward this `CompletableFuture` as a result. In case of global fail-over, the result would only complete if the discard call terminated.

The `CompletableFuture` returned by the `StopWithSavepointTerminationManager` could be collected by the `SchedulerBase` now. Multiple calls to `stopWithSavepoint` would result in combining the `CompletableFuture` results into a single `CompletableFuture`. We could then wait in `SchedulerBase.shutDownCheckpointServices` for this `CompletableFuture` to finish. It gets executed when the termination future of the underlying `ExecutionGraph` terminates. The drawback of this approach is that we would block in the main executor which might go against the rule of not having blocking operations on the main executor.
The solution would be to introduce a terminationFuture that only completes if the executions are done and is then handled in `JobMaster.stopScheduling` which is called by `JobMaster.stopJobExecution`.
It felt like it's getting a bit out of scope for the issue here - that's why we decided to go for the graceful shutdown approach for now and discuss the solution first.

tillrohrmann replied:
Yes, to do it properly would mean a lot more work. What we would have to do is to make the `SchedulerNG` asynchronously stoppable. One could for example let it extend `AutoCloseableAsync`. The `JobMaster` would then have to close the scheduler asynchronously by calling the `closeAsync` method and then make its shutdown dependent on the returned future. The future itself should only be completed once all savepoints and potentially other operations have completed. That way we never have to execute anything blocking in the main thread.

Hence, introducing a grace period for shutting down the `ioExecutor` sounds like a good first step.

XComp replied:
We switched back to not removing the Savepoint in case of fail over due to the reasons mentioned in [this comment](https://github.com/apache/flink/pull/14847#issuecomment-785303451).

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,297 @@
+                        log.warn(
+                                "Error occurred while cleaning up completed savepoint due to stop-with-savepoint failure.",
+                                e);
+                        inconsistentFinalStateException.addSuppressed(e);
```
I understand your motives here. I am just wondering whether the non-determinism due to the concurrency is a problem.

XComp replied:
Good point, I haven't considered that. I am going to add the savepoint's location to the log message to enable the user to delete the savepoint manually.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationManager.java:
```diff
@@ -0,0 +1,78 @@
+     *     operations run on.
+     * @return A {@code CompletableFuture} containing the path to the created savepoint.
+     */
+    public CompletableFuture<String> stopWithSavepoint(
+            CompletableFuture<CompletedCheckpoint> completedSavepointFuture,
+            CompletableFuture<Collection<ExecutionState>> terminatedExecutionStatesFuture,
+            ComponentMainThreadExecutor mainThreadExecutor) {
+        FutureUtils.assertNoException(
+                completedSavepointFuture
+                        // the completedSavepointFuture could also be completed by
+                        // CheckpointCanceller which doesn't run in the mainThreadExecutor
+                        .handleAsync(
+                                (completedSavepoint, throwable) -> {
+                                    stopWithSavepointTerminationHandler.handleSavepointCreation(
+                                            completedSavepoint, throwable);
+                                    return null;
+                                },
+                                mainThreadExecutor)
+                        .thenRun(
+                                () ->
+                                        FutureUtils.assertNoException(
+                                                terminatedExecutionStatesFuture.thenAccept(
+                                                        stopWithSavepointTerminationHandler
+                                                                ::handleExecutionsTermination))));
+
+        return stopWithSavepointTerminationHandler.getSavepointPath();
+    }
```
Nice :-)

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationManagerTest.java:
```diff
@@ -0,0 +1,192 @@
+ * StopWithSavepointTerminationHandler} regardless of the completion of the provided {@code
+ * CompletableFutures}.
+ */
+public class StopWithSavepointTerminationManagerTest extends TestLogger {
```
Nice, I like the tests. They are very clean now :-)

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImplTest.java:
```diff
@@ -0,0 +1,202 @@
+
+        // the checkpoint scheduling should be enabled in case of failure
+        assertTrue("Checkpoint scheduling should be enabled.", checkpointScheduling.isEnabled());
+    }
```
Maybe add one more tests which ensures that we can call `handleExecutionsTermination` after a failed savepoint and that this is ignored.

XComp replied:
Good point. I added two more test cases:
1. Checks that no global failover happens if the execution terminates with a failure.
2. Checks that checkpoint scheduling is not disabled if the execution terminates successfully.

tillrohrmann reviewed (commented):
Thanks for updating this PR @XComp. The changes look really good. I had a couple of smaller comments and questions. Please take a look.

XComp commented:
Thanks for looking into the changes, @tillrohrmann. I addressed your comments and replied to some of them. Please have a look.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -835,8 +840,21 @@ public void updateAccumulators(final AccumulatorSnapshot accumulatorSnapshot) {
+        final CheckpointCoordinator checkpointCoordinator = getCheckpointCoordinator();
+        Preconditions.checkState(
+                checkpointCoordinator != null,
+                "Checkpointing cannot be started since it's not enabled.");
```
The checkpoint scheduler can be null, in situation I consider valid:
scenario:
- job with no restarts configured
- savepoint creation fails

- result: the EG reaches a terminal state, and sets the checkpoint coordinator to null. This precondition fails, killing the JobManager process.

XComp replied:
I agree: That's a valid scenario. I'm just wondering now whether we should simply make the call optional in that case. We could just add another condition to the if statement. I don't like that because it makes the interface less specific. On the other hand one could argue that starting the checkpoint scheduling was always optional due to the condition that checks whether periodic checkpointing is enabled.

Handling it We cannot just wait for the termination to restart the scheduling due to the happy path. Another alternative would be to extend the interface by a method to check the state of the checkpointing (like `CheckpointCoordinator.shutdown` and only re-enable it if it's not shutdown. @rmetzger any thoughts on this.

XComp replied:
I went for the first option. Unfortunately, I had to remove the `CheckpointScheduling` interface from the `CheckpointCoordinator` due to that since the `CheckpointCoordinator` is throwing an exception in case of a shutdown.

The approach with having a method for checking whether Checkpoint Scheduling is enabled wasn't better either because it would have meant: Adding this check would have exposed the shutdown behavior of the `CheckpointCoordinator` which, as a consequence, would have required to add the `shutdown` method to the interface as well. This method doesn't work for the Scheduler, though. Hence, I decided to go for the approach that checks the availability of the `CheckpointCoordinator` internally.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,295 @@
+        state = state.onSavepointCreation(completedCheckpoint);
+
+        log.debug(
+                "Stop-with-savepoint transitioned from {} to {} on savepoint creation handling for job {}.",
```
Maybe it makes sense to implement State.toString() for nicer log messages:
```
4851 [flink-akka.actor.default-dispatcher-2] WARN  org.apache.flink.runtime.scheduler.adaptive.AdaptiveScheduler [] - Stop-with-savepoint transitioned from org.apache.flink.runtime.scheduler.stopwithsavepoint.StopWithSavepointTerminationHandlerImpl$FinalState@666f12e8 to org.apache.flink.runtime.scheduler.stopwithsavepoint.StopWithSavepointTerminationHandlerImpl$FinalState@666f12e8 on execution termination handling for job 6d1aa657a520797ad9f4070d6a6878d8 with some executions being in an not-finished state: [CANCELED, FAILED]
```

XComp replied:
Great finding 👍 Thanks

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -541,6 +544,59 @@ public void handleGlobalFailure() {
+     * updates.
+     */
+    @Test
+    public void handleGlobalFailureWithLocalFailure() {
```
We cannot remove [assert in SchedulerBase.restoreState](https://github.com/XComp/flink/blob/7cbd97f815d3bfd4715ddd8cb1d88f92f05a282a/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java#L419) as the following operation relies on all vertices being passed for restoring the state. Accepting less `ExecutionJobVertices` would mean that not all tasks would get restored.

I added this test to verify the case where a local task failure proceeds a global fail-over before actually restarting the tasks. The local task failure would not trigger a version update because of the [isNotifiable](https://github.com/XComp/flink/blob/7cbd97f815d3bfd4715ddd8cb1d88f92f05a282a/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java#L654) check in `SchedulerBase.updateTaskExecutionState`. This will only return `true` (and trigger the internal task state change) if the corresponding `ExecutionVertex` is either in state `FINISHED` or `FAILED`. The previous global fail-over has switched the state already to `CANCELING`. The local task failure will finally switch the state to `CANCELED`. Hence, no internal task state change is processed for the local task failure. No version upgrade is triggered and the global fail-over restart restarts the all vertices as expected.

tillrohrmann commented:
I think I have given bad advice on the savepoint clean up @XComp. I think that the `CheckpointCoordinator` will use the savepoint for recovery once it has been completed. Hence, we actually cannot delete the savepoint if a failure happens after the successful creation of the savepoint. Sorry for misguiding you here.

The only solution I can think of atm is that we could log the path to the savepoint so that the user can manually clean it up after the job has been terminated. I think we can only properly solve this problem after [FLIP-47](https://cwiki.apache.org/confluence/display/FLINK/FLIP-47%3A+Checkpoints+vs.+Savepoints) has been implemented.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImplTest.java:
```diff
@@ -0,0 +1,228 @@
+
+/**
+ * {@code StopWithSavepointTerminationHandlerImplTest} tests the stop-with-savepoint functionality
+ * of {@link SchedulerBase#stopWithSavepoint(String, boolean)}.
```
Don't we test the `StopWithSavepointTerminationHandlerImpl` here?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImplTest.java:
```diff
@@ -0,0 +1,228 @@
+
+        assertTrue("Savepoint should be discarded.", streamStateHandle.isDisposed());
+
+        // the checkpoint scheduling should be enabled in case of failure
+        assertTrue("Checkpoint scheduling should be enabled.", checkpointScheduling.isEnabled());
```
In case of a global failover, the checkpointing should actually be stopped for the time until the job has been restarted. Normally the de-/activation should happen via `CheckpointCoordinatorDeActivator`. In the current code what happens is that by calling the global failover the job will go to `RESTARTING` and then we go into `StopWithSavepointTerminationHandlerImpl.terminateExceptionally` and restart the checkpoint scheduling.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImpl.java:
```diff
@@ -0,0 +1,300 @@
+        FlinkException inconsistentFinalStateException = new FlinkException(errorMessage);
+
+        scheduler.handleGlobalFailure(inconsistentFinalStateException);
+        discardSavepoint(completedSavepoint);
```
I think we must not discard the savepoint because it might be used by Flink for recoveries. Sorry for the bad advice.

tillrohrmann reviewed (changes requested):
Thanks for updating the PR @XComp. I think it looks really good. Due to my wrong advice we need to update one thing with discarding the savepoint.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointContext.java:
```diff
@@ -0,0 +1,197 @@
+    }
+
+    private StopWithSavepointState terminateExceptionally(Throwable throwable) {
+        scheduler.startCheckpointScheduler(checkpointCoordinator);
```
This issue was addressed in @rmetzger [comment](https://github.com/apache/flink/pull/14847#discussion_r581918367)

XComp commented:
I squashed and rebased the branch adding another cleanup hotfix.
@AHeise Could you double-check your change [5bf7da7](https://github.com/apache/flink/pull/14847/commits/5bf7da77c3052f12197075c6a221ef0815377e8a). I had to resolve some minor conflicts due to @pnowojski fix [19ceee05](https://github.com/apache/flink/commit/19ceee05)

XComp commented:
I rebased once more due to `UnalignedCheckpointCompatibilityITCase` failing: The test was fixed on `master` by [6d2a05a](https://github.com/apache/flink/commit/6d2a05a72fc73c60473078a050b4ea4f79b61a27).

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -883,10 +887,26 @@ public void updateAccumulators(final AccumulatorSnapshot accumulatorSnapshot) {
+            log.info(
+                    "Periodic checkpoint scheduling could not be stopped due to the CheckpointCoordinator being shutdown.");
+        } else {
+            getCheckpointCoordinator().stopCheckpointScheduler();
```
```suggestion
           checkpointCoordinator.stopCheckpointScheduler();
```

AHeise commented:
> I squashed and rebased the branch adding another cleanup hotfix.
> @AHeise Could you double-check your change [c249642](https://github.com/apache/flink/pull/14847/commits/c249642b7a63037ad03de2a936aeee6420129a0b). I had to resolve some minor conflicts due to @pnowojski fix [19ceee05](https://github.com/apache/flink/commit/19ceee05)

LGTM!

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointTerminationHandlerImplTest.java:
```diff
@@ -0,0 +1,226 @@
+
+        assertFalse("Savepoint should not be discarded.", streamStateHandle.isDisposed());
+
+        // the checkpoint scheduling should be enabled in case of failure
```
```suggestion
        // the checkpoint scheduling should not be enabled in case of failure
```

tillrohrmann reviewed (approved):
Thanks for creating this PR @XComp. The changes look good to me. I had two very minor comments. +1 for merging this PR once AZP gives green light. Could you also prepare PRs against `release-1.12` and `release-1.11`?

XComp commented:
Thanks @tillrohrmann for the remarks. I applied your comments and created PR #15034 (`release-1.12`) and PR #15035 (`release-1.11`).

XComp commented:
[Build #290](https://dev.azure.com/mapohl/flink/_build/results?buildId=290&view=results) contains two failures:
* [FileBufferReaderITCase.testSequentialReading](https://dev.azure.com/mapohl/flink/_build/results?buildId=290&view=logs&j=9dc1b5dc-bcfa-5f83-eaa7-0cb181ddc267&t=ab910030-93db-52a7-74a3-34a0addb481b) is covered by [FLINK-21416](https://issues.apache.org/jira/browse/FLINK-21416)
* I created [FLINK-21515](https://issues.apache.org/jira/browse/FLINK-21515) to cover [SourceStreamTaskTest.testStopWithSavepointShouldNotInterruptTheSource](https://dev.azure.com/mapohl/flink/_build/results?buildId=290&view=logs&j=cc649950-03e9-5fae-8326-2f1ad744b536&t=51cab6ca-669f-5dc0-221d-1e4f7dc4fc85&l=8089)

tillrohrmann commented:
The test failures seem to be unrelated. Also the e2e have passed for `93304d6`. Merging it now.
